src/main/java/gregtech/api/metatileentity/multiblock/MultiblockControllerBase.java

```diff
@@ -333,64 +437,229 @@ protected Function<BlockPos, Integer> multiblockPartSorter() {
+        List<BlockPos> pos = new ArrayList<>();
+        for (Long2ObjectMap.Entry<BlockInfo> entry : cache.long2ObjectEntrySet()) {
+            if (entry.getValue().getBlockState().getBlock() instanceof VariantActiveBlock<?>) {
+                pos.add(BlockPos.fromLong(entry.getLongKey()));
             }
         }
+        return pos;
+    }
+
+    protected void registerMultiblockAbility(IMultiblockAbilityPart<Object> part) {
+        List<Object> abilityList = multiblockAbilities.computeIfAbsent(part.getAbility(), k -> new ArrayList<>());
+        part.registerAbilities(abilityList);
+    }
+
+    protected void forEachFormed(String name, BiConsumer<BlockInfo, GreggyBlockPos> action) {
+        Long2ObjectMap<BlockInfo> cache = getSubstructure(name).getCache();
+        GreggyBlockPos pos = new GreggyBlockPos();
+        for (Long2ObjectMap.Entry<BlockInfo> entry : cache.long2ObjectEntrySet()) {
+            action.accept(entry.getValue(), pos.fromLong(entry.getLongKey()));
+        }
+    }
+
+    protected void formStructure(String name) {
+        getSubstructure(name).getPatternState().setFormed(true);
+        setFlipped(getSubstructure(name).getPatternState().isFlipped(), name);
+        writeCustomData(STRUCTURE_FORMED, buf -> buf.writeString(name).writeBoolean(true));
     }
 
-    protected void formStructure(PatternMatchContext context) {}
+    protected void formStructure() {
+        formStructure("MAIN");
+    }
 
     public void invalidateStructure() {
-        this.multiblockParts.forEach(part -> part.removeFromMultiBlock(this));
-        this.multiblockAbilities.clear();
-        this.multiblockParts.clear();
-        this.structureFormed = false;
-        this.setFlipped(false);
-        writeCustomData(STRUCTURE_FORMED, buf -> buf.writeBoolean(false));
+        invalidateStructure("MAIN");
+    }
+
+    public void invalidateStructure(String name) {
+        if (!getSubstructure(name).getPatternState().isFormed()) return;
+        // i am sorry
+        Object[] added = { null };
+        List<Object> dummyList = new ArrayList<>() {
+
+            @Override
+            public boolean add(Object e) {
+                added[0] = e;
+                return true;
+            }
+        };
+
+        multiblockParts.removeIf(part -> {
+            if (name.equals(part.getSubstructureName())) {
+                if (part instanceof IMultiblockAbilityPart<?>) {
+                    // noinspection unchecked
+                    IMultiblockAbilityPart<Object> ability = (IMultiblockAbilityPart<Object>) part;
+                    added[0] = null;
+                    ability.registerAbilities(dummyList);
+                    if (added[0] != null) multiblockAbilities.get(ability.getAbility()).remove(added[0]);
+                }
+                part.removeFromMultiBlock(this);
+                return true;
+            }
+            return false;
+        });
```

Comment:
an ability part can add several abilities to the multiblock (see quad/nonuple fluid hatches), but this only removes the last ability added. i think you can just use a proper list here as well.